plugins/woocommerce/src/Database/Migrations/MigrationHelper.php

```diff
@@ -75,4 +77,163 @@ public static function generate_on_duplicate_statement_clause( array $columns ):
+	 * @param array  $old_to_new_states_mapping An associative array where keys are the old state codes and values are the new state codes.
+	 * @return void
+	 */
+	private static function migrate_country_states_for_cot_orders( string $country_code, array $old_to_new_states_mapping ): void {
+		global $wpdb;
+
+		$current_date_gmt          = current_time( 'Y-m-d H:i:s', true );
+		$states_as_comma_separated = "('" . join( "','", array_keys( $old_to_new_states_mapping ) ) . "')";
+
+		// phpcs:disable WordPress.DB.PreparedSQL.NotPrepared, WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+
+		$sql            = $wpdb->prepare(
+			"SELECT id, order_id, state FROM {$wpdb->prefix}wc_order_addresses WHERE country=%s AND state IN {$states_as_comma_separated}",
+			$country_code
+		);
+		$addresses_data = $wpdb->get_results( $sql, ARRAY_A );
+
+		foreach ( $addresses_data as $address_data ) {
+			$update_query = $wpdb->prepare(
+				"UPDATE {$wpdb->prefix}wc_order_addresses SET state=%s WHERE id=%d",
+				$old_to_new_states_mapping[ $address_data['state'] ],
+				$address_data['id']
+			);
+			$wpdb->query( $update_query );
+
+			$update_query = $wpdb->prepare(
+				"UPDATE {$wpdb->prefix}wc_orders SET date_updated_gmt=%s WHERE id=%d",
+				$current_date_gmt,
+				$address_data['order_id']
+			);
+			$wpdb->query( $update_query );
+		}
+
+		// phpcs:enable WordPress.DB.PreparedSQL.NotPrepared, WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+	}
+
+	/**
+	 * Migrate state codes for orders in the posts table.
+	 *
+	 * @param string $country_code The country that has the states for which the migration is needed.
+	 * @param array  $old_to_new_states_mapping An associative array where keys are the old state codes and values are the new state codes.
+	 * @return void
+	 */
+	private static function migrate_country_states_for_post_orders( string $country_code, array $old_to_new_states_mapping ): void {
+		global $wpdb;
+
+		$current_date              = current_time( 'Y-m-d H:i:s', false );
+		$current_date_gmt          = current_time( 'Y-m-d H:i:s', true );
+		$states_as_comma_separated = "('" . join( "','", array_keys( $old_to_new_states_mapping ) ) . "')";
+
+		// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+		$sql = $wpdb->prepare(
+			"SELECT meta_id, post_id, meta_value FROM {$wpdb->prefix}postmeta
+				WHERE (meta_key='_billing_state' OR meta_key='_shipping_state')
+				AND meta_value IN {$states_as_comma_separated}
+				AND post_id IN (
+					SELECT post_id FROM {$wpdb->prefix}postmeta WHERE
+					(meta_key = '_billing_country' OR meta_key='_shipping_country')
+					AND meta_value=%s
+				)",
+			$country_code
+		);
+		// phpcs:enable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+
+		// phpcs:disable WordPress.DB.PreparedSQL.NotPrepared
+		$addresses_data = $wpdb->get_results( $sql, ARRAY_A );
+
+		foreach ( $addresses_data as $address_data ) {
```

Comment:
If I understand correctly, this will run one SELECT and `2n` UPDATEs where `n` = the number of affected orders.

I wonder if it would be possible to run `c` SELECT and `c` UPDATES where `c` = the number of states within a country. Each UPDATE would update the meta_value of all rows for the same state (across all orders). I think `c << n` most of the time, so it seems more favorable. 

Also, have you tested how this behaves if the number of affected orders is large (e.g. 1m)? I guess it will time out and fail, right?

Reply:
Again, no longer relevant after the recent code changes (I'm no longer querying the orders/posts tables, only the meta tables).